invoices: add client to map before delivering backlog notif

Crypt-iQ opened the pull request:
Prior to this change, if SubscribeSingleInvoice was called, it was possible that a state change would never be delivered to the client. The sequence of events would be:
- deliverSingleBacklogEvents called with invoice in Open state
- invoice goes to the Accepted state, no client to notify
- client added to map

This is fixed by adding the client to the map first. However, with this change it then becomes possible (though perhaps unlikely) that the client receives an Accepted and then an Open notification since the notifications occur in different goroutines. Whether this is an acceptable tradeoff is an Open question

Fixes https://github.com/lightningnetwork/lnd/issues/6475

yyforyongyu commented:
I've seen this flake before, and whether to fix it in the test or in lnd is probably an open question. I previously fixed it by subscribing early in the test so no events were missed. I think subscriptions need to be made as early as possible so the client won't miss an event.

What you described here is different tho, that the subscription is missing events internally. Maybe we need to add the lock back where it was deleted by me previously🙈

> with this change it then becomes possible (though perhaps unlikely) that the client receives an Accepted and then an Open notification since the notifications occur in different goroutines. 

I've seen this before while working on #6053.

Crypt-iQ commented:
I think this can also happen with `SubscribeNotifications`. It seems a bit harder to address the issue of out-of-order events in that case since there may be multiple events in the backlog.

This out-of-order event problem can be fixed in the `SubscribeSingleInvoice` case by using a mutex and updating a boolean when calling `client.notify`. I don't think only using an atomic variable would work since according to https://dave.cheney.net/2015/08/08/performance-without-the-event-loop under the Goroutines section, there are four cases a goroutine can yield to the scheduler:
- Channel send and receive operations, if those operations would block.
- The go statement, although there is no guarantee that new goroutine will be scheduled immediately.
- Blocking syscalls like file and network operations.
- After being stopped for a garbage collection cycle.

It seems like only the fourth case matters here. I'm not at all familiar with how the OS or Go handles threads (and by extension goroutines), but it seems to me that _only_ checking the value of an atomic doesn't work if one goroutine is paused and another one resumes.

The easiest fix for the itest would be to move the subscribe call earlier, but that doesn't address missed events

Crypt-iQ commented:
Changed `SubscribeNotifications` as well. I added a commit that should prevent a possible case of hanging on delivery to the `ntfnQueue`. Also added a commit that should fix the OOO delivery of updates by ensuring the backlog gets to the `ntfnQueue` first. Can make a more detailed comment tomorrow

yyforyongyu reviewed (commented):
Nice fix! Pending a release note, otherwise LGTM👍

Crypt-iQ commented:
Prior to this PR, I believe it was possible that (though I couldn't test it) using the `client.canceled` atomic to guard delivery would not work as intended:
```
goroutine 1

* t1. dispatchToSingleClients
   * t2. compare client.canceled against 1 (it's 0)
   * t3. proceed to client.notify(...) which sends along the ntfnQueue
```

```
goroutine 2

* t2.x (sometime between t2 and t3) - client.Cancel() called, which sets client.canceled to 1
   * t2.x (sometime between t2 and t3) - i.ntfnQueue.Stop called, which stops the queue from listening on the in channel
```

The result would be that at time t3, notify would hang since the ntfnQueue is stopped and the invoiceregistry isn't shutting down. This is fixed by listening on the `client.cancelChan` which is closed when `client.Cancel()` has completed (the waitgroup is totally unused and I'll also remove that).

The second change is that the `backlogDelivered` chan is closed when backlog events in `SubscribeSingleInvoice` or `SubscribeNotifications` have been delivered to the ntfnQueue. Since the closing is in a defer statement, this chan will always be closed.

yyforyongyu commented:
Nice catch on the `canceled` issue! Looks like there's a unit test failing, not sure if it's related but did see it's invoice registry related.

Crypt-iQ commented:
> Nice catch on the `canceled` issue! Looks like there's a unit test failing, not sure if it's related but did see it's invoice registry related.

it's testchannellinkcancelfullcommitment which has a deadlock

yyforyongyu reviewed (approved):
LGTM👍

Roasbeef commented on invoices/invoiceregistry.go:
```diff
@@ -355,12 +355,14 @@ func (i *InvoiceRegistry) dispatchToSingleClients(event *invoiceEvent) {
+		case <-client.backlogDelivered:
+			// We won't deliver any events until the backlog has
+			// went through first.
+		case <-i.quit:
```
Shouldn't we still be listening on the client's quit channel here? As otherwise the client may not be there at all, but then we add more notifications to its queue.

Roasbeef commented on invoices/invoiceregistry.go:
```diff
@@ -1452,6 +1467,10 @@ func (i *invoiceSubscriptionKit) Cancel() {
 func (i *invoiceSubscriptionKit) notify(event *invoiceEvent) error {
 	select {
 	case i.ntfnQueue.ChanIn() <- event:
+	case <-i.cancelChan:
```
Ah ok, I see the move to start listening on the cancel channel here instead.

Roasbeef commented on invoices/invoiceregistry.go:
```diff
@@ -1567,14 +1590,18 @@ func (i *InvoiceRegistry) SubscribeSingleInvoice(
 	client.ntfnQueue.Start()
 
+	// This notifies other goroutines that the backlog phase is done.
+	defer close(client.backlogDelivered)
```
👍

Roasbeef reviewed (approved):
LGTM 🧜‍♂️

Excellent diagnosis, reproduction scenario and fix!

Roasbeef commented:
Just needs a rebase!

Crypt-iQ commented:
rebased
